### PR TITLE
Prerender non-index pages

### PR DESCRIPTION
**Pull request by macovedj**
This PR takes care of what is brought up [here](https://github.com/redwoodjs/redwood/issues/3110)

### CODE REVIEW

**Comment by dthyresson** [comment]
Hi @macovedj and thanks for the PR.

Could you provide a description above to let us know why non-index pages should also be prerendered and also perhaps an example or user case supporting this change?

That will help us review the PR.

Thanks!

**Comment by macovedj** [comment]
@dthyresson I linked the issue that inspired this PR.  In general, I think prerendering offers benefits for both SEO and performance

**Comment by dac09** [comment]
Hey @macovedj thanks for this. Few things that need to be addressed I think:

- [ ] Use prettier to format the code (I recommend the vscode prettier extension)
- [ ] Some minor comments in the PR
- [ ] If path at page isn't found, we should render 200.html (if present), or index.html (if 200 not present)

And from the original checklist in the issue:
- [x] Prerender top level page e.g. site.com/about should serve `web/dist/about.html`
- [ ] Prerender nested page i.e. trailing slashes not needed e.g. site.com/dashboard should serve `web/dist/dashboard/index.html` (not sure)
- [ ] Prerender nested named page e.g. site.com/dashboard/otherpage should serve `web/dist/dashboard/otherpage.html`
- [x] Handle trailing slashes e.g. site.com/about/ should serve `web/dist/about.html`

File: packages/api-server/src/middleware/withWebServer.ts [hunk 1]
```diff
@@ -6,11 +6,21 @@ import type { Application } from 'express'
...
+type Contents = {
+  [content: string]: string
+}
 const withWebServer = (app: Application) => {
-  const indexContent = fs.readFileSync(
-    path.join(getPaths().web.dist, '/index.html'),
-    'utf-8'
-  )
+  const files = fs.readdirSync(getPaths().web.dist)
+    .filter(file => file.split('.')[1] === 'html')
```
Comment (dac09) [inline packages/api-server/src/middleware/withWebServer.ts:14]: do you think we could use `path.extname`? https://nodejs.org/api/path.html#path_path_extname_path

File: packages/api-server/src/middleware/withWebServer.ts [hunk 2]
```diff
@@ -6,11 +6,21 @@ import type { Application } from 'express'
...
+    .map(file => file.split('.')[0])
```
Comment (dac09) [inline packages/api-server/src/middleware/withWebServer.ts:15]: Could we use `path.basename` here? https://nodejs.org/api/path.html#path_path_basename_path_ext

File: packages/api-server/src/middleware/withWebServer.ts [hunk 3]
```diff
@@ -6,11 +6,21 @@ import type { Application } from 'express'
 
 import { getPaths } from '@redwoodjs/internal'
 
+type Contents = {
+  [content: string]: string
```
Comment (dac09) [inline packages/api-server/src/middleware/withWebServer.ts:10]:
```suggestion
  [path: string]: string
```

File: packages/api-server/src/middleware/withWebServer.ts [hunk 4]
```diff
@@ -19,8 +29,10 @@ const withWebServer = (app: Application) => {
   )
 
   // For SPA routing on unmatched routes
-  app.get('*', function (_, response) {
-    response.send(indexContent)
+  Object.keys(contents).forEach((content) => {
+    app.get(`/${content}`, function (_, response) {
+      response.send(contents[content])
```
Comment (dac09) [inline packages/api-server/src/middleware/withWebServer.ts:34]: Do you think its worth setting the content type header here?

File: packages/api-server/src/middleware/withWebServer.ts [hunk 5]
```diff
@@ -6,11 +6,21 @@ import type { Application } from 'express'
...
 const withWebServer = (app: Application) => {
-  const indexContent = fs.readFileSync(
-    path.join(getPaths().web.dist, '/index.html'),
-    'utf-8'
-  )
+  const files = fs.readdirSync(getPaths().web.dist)
+    .filter(file => file.split('.')[1] === 'html')
+    .map(file => file.split('.')[0])
+
+  const contents: Contents = files.reduce((acc, cur) => ({
```
Comment (dac09) [inline packages/api-server/src/middleware/withWebServer.ts:17]: Suggested name: htmlContentsByPath

File: packages/api-server/src/middleware/withWebServer.ts [hunk 6]
```diff
@@ -6,11 +6,21 @@ import type { Application } from 'express'
 
 import { getPaths } from '@redwoodjs/internal'
 
+type Contents = {
```
Comment (dac09) [inline packages/api-server/src/middleware/withWebServer.ts:9]:
```suggestion
type HtmlContents = {
```

**Review by dac09 — CHANGES_REQUESTED** [review]

File: packages/api-server/src/middleware/withWebServer.ts (hunk 3 above)
Comment (macovedj) [reply packages/api-server/src/middleware/withWebServer.ts:10]: I struggled with the name here, but path doesn't feel much better to me than content.  The values in the object that gets declared using this type are actually the html for the corresponding paths.  Does that impact what you think it should be called?

File: packages/api-server/src/middleware/withWebServer.ts (hunk 4 above)
Comment (macovedj) [reply packages/api-server/src/middleware/withWebServer.ts]: I'm conflicted about what the value should be for prerendering a javascript application with html.  Do you have opinions?

Comment (dac09) [reply packages/api-server/src/middleware/withWebServer.ts]: I think its just plain html :) `Content-Type: text/html; charset=UTF-8`

**Comment by macovedj** [comment]
@dac09 I'm going to use [this](https://github.com/isaacs/node-glob) for nested routes... unless you're opinionated about doing it another way.

**Comment by macovedj** [comment]
> @dac09 I'm going to use [this](https://github.com/isaacs/node-glob) for nested routes... unless you're opinionated about doing it another way.

ehh, actually comparing a few options

**Comment by dac09** [comment]
@macovedj an idea might be to implement a function in `@redwoodjs/internal` like getApiServerFunctions (you can use the same library)

**Comment by macovedj** [comment]
@dac09 prerendering nested routes now.  Have some questions re: under what circumstances empty directories would be built, as well as directories with only an index.html, as well ass if 200.html files and 404.html files already exist anywhere.

**Comment by dac09** [comment]
@macovedj nice one! I've pushed up some changes for the remaining item. I think its looking pretty good :), thanks for all the effort into this!

**Review by dac09 — APPROVED** [review]

**Pull request merged**